community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/command/PhysicalLogCommandReaderV3_0.java

```diff
@@ -0,0 +1,850 @@
+        return false;
+    }
+
+    @Override
+    public boolean visitIndexRemoveCommand( RemoveCommand command ) throws IOException
+    {
+        IndexCommandHeader header = readIndexCommandHeader();
+        Number entityId = header.entityIdNeedsLong ? channel.getLong() : channel.getInt();
+        Object value = readIndexValue( header.valueType );
+        command.init( header.indexNameId, header.entityType, entityId.longValue(), header.keyId, value );
+        return false;
+    }
+
+    @Override
+    public boolean visitIndexDeleteCommand( DeleteCommand deleteCommand ) throws IOException
+    {
+        IndexCommandHeader header = readIndexCommandHeader();
+        deleteCommand.init( header.indexNameId, header.entityType );
+        return false;
+    }
+
+    @Override
+    public boolean visitIndexCreateCommand( CreateCommand createCommand ) throws IOException
+    {
+        IndexCommandHeader header = readIndexCommandHeader();
+        Map<String,String> config = read2bMap( channel );
+        createCommand.init( header.indexNameId, header.entityType, config );
+        return false;
+    }
+
+    @Override
+    public boolean visitIndexDefineCommand( IndexDefineCommand indexDefineCommand ) throws IOException
+    {
+        readIndexCommandHeader();
+        Map<String,Integer> indexNames = readMap( channel );
+        Map<String,Integer> keys = readMap( channel );
+        indexDefineCommand.init( indexNames, keys );
+        return false;
+    }
+
+    @Override
+    public boolean visitNodeCountsCommand( NodeCountsCommand command ) throws IOException
+    {
+        int labelId = channel.getInt();
+        long delta = channel.getLong();
+        command.init( labelId, delta );
+        return false;
+    }
+
+    @Override
+    public boolean visitRelationshipCountsCommand( RelationshipCountsCommand command ) throws IOException
+    {
+        int startLabelId = channel.getInt();
+        int typeId = channel.getInt();
+        int endLabelId = channel.getInt();
+        long delta = channel.getLong();
+        command.init( startLabelId, typeId, endLabelId, delta );
+        return false;
+    }
+
+    private Map<String,Integer> readMap( ReadableLogChannel channel ) throws IOException
+    {
+        byte size = channel.get();
+        Map<String,Integer> result = new HashMap<>();
+        for ( int i = 0; i < size; i++ )
+        {
+            String key = read2bLengthAndString( channel );
+            int id = getUnsignedShort( channel );
+            if ( key == null )
+            {
+                return null;
+            }
+            result.put( key, id );
+        }
+        return result;
+    }
+
+    private int getUnsignedShort( ReadableLogChannel channel ) throws IOException
+    {
+        int result = channel.getShort() & 0xFFFF;
+        return result == 0xFFFF ? -1 : result;
+    }
+
+    private IndexCommandHeader readIndexCommandHeader() throws ReadPastEndException, IOException
```

Comment:
`ReadPastEndException` is a subclass of `IOException`. We can omit declaring it.